### PR TITLE
Symbols overlay

### LINKED ISSUE #1550: Add new constraint symbols
This is a [figma](https://www.figma.com/file/5UUPOawPLEs3D0Gptf42vy/Designs?type=design&node-id=1260%3A3441&mode=design&t=83F46AJ6NA0G1ThW-1) grab

Design is not finalised so is likely to become stale.

The idea is we communicate how each segment is defined with symbols, and make them bolder or bright if the input is constrained (non-literal). We will also communicate how segments relate to each other with a little dot above the symbol, by highlighting the source-ranges and highlighting the related segments.

image

image

### PR DESCRIPTION
**Pull request by Irev-Dev**
closes #1550 

WIP screenshot
image

### Here are the broad strokes of how the overlays work.
- [updateSegment](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-99554257001e5edcf12884f13e5b8abad4cda74888e139c75198965dba220c3aL1016) has been updated to return a function that has information about each segment
- it returns a function so that they can all be called together later so to not cause perf issues when a segment is being dragged (because this calls updateSegment very often)
- The information for the segment that we get from this is [basically this here](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-5b3927b16824249515f7aa8762bd6c9c9ba641a023bb6d0a75b75331326cf4aaR97). There's a few things there, but most importantly is the windowCoords, as this is how we know where to position all of the coordinates.
- The math for how we [get these windowCoords is here](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-e622a7de7edc2476f4f3b0e2f8e038e5c4c4c01c484e79cb10851bf2e2577682R198-R241), but the basics of it is that given a three.js entity (like the segment arrowheads) we can project that into screen space.
- The info for all the segments get added to xState context, from there we can render them like any other react component.
- All of the react component stuff is in [ClientSideSceneComp](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-68cf6c5f4c0dc2b8b7833730e89b3e4477de4d1392e55785f10a84621fed727c), the math of exactly where we place the components [is here](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-68cf6c5f4c0dc2b8b7833730e89b3e4477de4d1392e55785f10a84621fed727cR197-R226) because there's some nuances about exactly where we put the overlay depending on the segments angle and what not.
- Are far as overlay info goes we've only covered how we know where to place them and not the specifics that's going to depend on the type of segment `line`, `angledLien`, `tangentialArcTo` etc. because this is going to change depending on the segment it's best to co-locate these with other information about these stdlib functions.
- The way this is done is by adding the `getConstraintInfo` function type to [stdTypes](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-87f4f99856ae388f40aad681717c84929ed617de4d37d7f71571405146e47cacR130-R134)
- This now requires these function to defined for each stdlib function in [sketch.ts](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-542f7543970ec85c9113fb0271d0bf54c3826b956e0a2a7348e7548330f39cb8) In that file you can see they are all defined but with some abstractions since there are similarities between each.
- The[ information we get from this](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-542f7543970ec85c9113fb0271d0bf54c3826b956e0a2a7348e7548330f39cb8) is the name of the stdlib function, what kind of input it is `xRelative`, `angle` etc, the `pathToNode` and more.
- All of these individual `getConstraintInfo` is abstracted to the master [getConstraintInfo](https://github.com/KittyCAD/modeling-app/pull/2033/files#diff-542f7543970ec85c9113fb0271d0bf54c3826b956e0a2a7348e7548330f39cb8R1559) that takes the callExpression node for a segment and returns an array of contraintInfo
- From there I'd say most of the rest of the diff in this PR is both tests, and ast query and mod cod for allowing the constraints to be added/removed by clicking the overlays as well as the delete segment button.

### TODO list
- [x] add delete to thee-dot menu
  - [x] make it safe (query for segments that depend on the segment being deleted) ([1025e66](https://github.com/KittyCAD/modeling-app/pull/2033/commits/1025e66fcccb3bea9d53d827e240775747c35065))
- [x] clean up overlays after sketch is exited ([43a3e6a](https://github.com/KittyCAD/modeling-app/pull/2033/commits/43a3e6a70329c3c730631677c3a5d390204cf466))
- [x] make sure overlays adjust properly on rejig
- [x] Make robust to small changes in source ranges (literal changing from 1.01 to -10.25 is two characters longer and can cause issues if we source ranges instead of pathToNode) ([8ad0452](https://github.com/KittyCAD/modeling-app/pull/2033/commits/8ad04521c245cb57f97db8d3caf1c0c6d69bedfc))
- [x] Make overlays only show on hover for a segment (but then stay for 500ms or so) ([e0a33b1](https://github.com/KittyCAD/modeling-app/pull/2033/commits/e0a33b10a278814831ecec5028f246c58047229f))
- [x] Overlays should assist tests, (therefore they need to exist when not visible to users) ([dcb19f0](https://github.com/KittyCAD/modeling-app/pull/2033/commits/dcb19f06d6bd6d641706fe79231d12fffdef3575))
  - [x] Refactor at least one tests to remove hardcoded px coords ([dcb19f0](https://github.com/KittyCAD/modeling-app/pull/2033/commits/dcb19f06d6bd6d641706fe79231d12fffdef3575))
- [x] overlay should not appear for very small (in screen space) segment, user will need to zoom in ([78a011d](https://github.com/KittyCAD/modeling-app/pull/2033/commits/78a011d8e73818c93ab25b9aaebc812e6065c7d4))
- [x] Clicking symbols should add or remove constraint ([7371cb2](https://github.com/KittyCAD/modeling-app/pull/2033/commits/7371cb2806f5c044b8150e5bef496e470c8516f9))
- [x] Make CI happy again.
- [x] Double-check performance ([5bd19c6](https://github.com/KittyCAD/modeling-app/pull/2033/commits/5bd19c6cbbdb815073265bd6acd0d73cd652f62e) yup found a lurking render issue that wasn't created in the PR but was made worse by)
- [x] Unit tests for:
   - [x] deleteSegmentFromPipeExpression ([6c6e4f7](https://github.com/KittyCAD/modeling-app/pull/2033/commits/6c6e4f7253eaf305a383df2a04ae25d879ad2e09))
   - [x] findUsesOfTagInPipe ([2b6a195](https://github.com/KittyCAD/modeling-app/pull/2033/commits/2b6a195541740a1f6eae0f61f43e8c629bf4fb95))
   - [x] removeSingleConstraintInfo/removeSingleConstraint ([eea199b](https://github.com/KittyCAD/modeling-app/pull/2033/commits/eea199b1dc8e81a114a663c21d5f5e6fba5e81df))
   - [x] getConstraintInfo ([ba51601](https://github.com/KittyCAD/modeling-app/pull/2033/commits/ba516012f719c3f86e7ac35d4e257d23849a44ad))
 - [x] e2e tests for:
   - [x] Hovering on segment for overlays
   - [x] Hovering on overlay to see popover and code highlight
   - [x] Clicking on Unconstrained overlay (add variable)
   - [x] Clicking on Constrained overlay (unconstrain it)
   - [x] Clicking on overlay three-dot and deleting segment ([f73e5d2](https://github.com/KittyCAD/modeling-app/pull/2033/commits/f73e5d2ffbab417339e999af009e3a79af78ba9d) - [4ae3b30](https://github.com/KittyCAD/modeling-app/pull/2033/commits/4ae3b30ff01b4b8c648801e753d605364b94ab4f))
     - [x] Deleting segment that have dependant segment
     - [x] Deleting segment that depends on the tag outside of the current sketch

Task for Frank:
- [x] Make overlays compatible with light mode

Fast follow tasks:
- [ ] https://github.com/KittyCAD/modeling-app/issues/2390
     
     
Other misc notes
- figure out why disabling button changes styles that can't be overridden
- implement dot above constraint (link of dependencies etc)

### CODE REVIEW

File: src/clientSideScene/ClientSideSceneComp.tsx [hunk 1]
```diff
@@ -100,17 +132,531 @@ export const ClientSideScene = ({
...
+              const transform = removeSingleConstraintInfo(
+                input,
+                kclManager.ast,
+                kclManager.programMemory
+              )
+              if (!transform) return
+              const { modifiedAst } = transform
+              kclManager.updateAst(modifiedAst, true)
+            } catch (e) {
+              console.log('error', e)
```
Comment (dshaw) [inline src/clientSideScene/ClientSideSceneComp.tsx:595]: Good to see this not throwing per #2325

**Review by dshaw — COMMENTED** [review]
I suggest having @jgomez720 run this through its paces and test the end user experience before landing.

I went through the code and it looks good. I got a lot of engine errors while doing an integration test. I'll sync up with @franknoirot to assess fully tomorrow.

File: src/index.tsx [hunk 2]
```diff
@@ -40,7 +40,11 @@ root.render(
             primary: 'oklch(89% 0.16 143.4deg)',
             secondary: 'oklch(48.62% 0.1654 142.5deg)',
           },
-          duration: 1500,
+          duration:
+            typeof window !== 'undefined' &&
+            window.localStorage.getItem('playwright') === 'true'
```
[the remark this answers is not in the document]
Comment (Irev-Dev) [reply src/index.tsx]: Sorry, an old habbit to make sure things are node friendly for unit tests, but unit tests are still happy.

You would know better than me

7829eeef3e376a33e1dfb9f2ab0207e963c7913e

**Comment by Irev-Dev** [comment]
> I went through the code and it looks good. I got a lot of engine errors while doing an integration test. I'll sync up with @franknoirot to assess fully tomorrow.

What are the engine error are these, I could look now if I knew.

**Comment by Irev-Dev** [comment]
I'm actually somewhat confused, we don't really have access to engine errors during the e2e tests as far as I'm aware, unless you're looking at some other dashboard i.e. not in the github action?

**Comment by dshaw** [comment]
> I'm actually somewhat confused, we don't really have access to engine errors during the e2e tests as far as I'm aware, unless you're looking at some other dashboard i.e. not in the github action?

Not e2e "integration test", but me running the code and integration testing on my system.

**Comment by Irev-Dev** [comment]
Ah I see.

File: src/clientSideScene/ClientSideSceneComp.tsx (hunk 1 above)
Comment (dshaw) [reply src/clientSideScene/ClientSideSceneComp.tsx:595]: `console.error` would be better here since it will show up as an error in console.

**Comment by dshaw** [comment]
Screenshot 2024-05-21 at 11 13 56 AM

Variables pane seems to have updated once and then stopped as I continued to add more.

**Comment by dshaw** [comment]
Re-ran unbuntu playwright tests that had timed out

**Comment by dshaw** [comment]
> What are the engine error are these, I could look now if I knew.

Did testing today on the [vertex instance](https://modeling-app-git-kurt-symbols-overlay-kittycad.vercel.app/) and was not seeing any engine errors. Might have been a local build issue. I've been working in rust and may not have properly reset my local environment to be able to test properly.

**Pull request merged**